File: com.bosch.fsp.master_of_plow.gen/src/com/bosch/nevonex/serial_ext_gps_nmea0183/impl/Serial_Ext_GPS_NMEA0183.java
/**
Copyright (c) Robert Bosch GmbH. All rights reserved.
*/
package com.bosch.nevonex.serial_ext_gps_nmea0183.impl;

import com.bosch.fsp.logger.FCALLogs;
import com.bosch.fsp.logger.LoggerConstants;

import com.bosch.fsp.platform.InterfaceDetails;

import com.bosch.fsp.runtime.feature.IMachine;

import com.bosch.nevonex.common.ITopicObject;
import com.bosch.nevonex.common.impl.TopicObject;

import com.bosch.nevonex.fcal.IGPS_NMEA0183_T_R;

import com.bosch.nevonex.serial_ext_gps_nmea0183.ISerial_Ext_GPS_NMEA0183;

import com.bosch.nevonex.types.IPropertyChange;
import com.bosch.nevonex.types.impl.TypesPackage;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import java.lang.reflect.InvocationTargetException;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.eclipse.emf.common.util.BasicEList;
import org.eclipse.emf.common.util.EList;

import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.InternalEObject;

import org.eclipse.emf.ecore.util.EDataTypeUniqueEList;

/**
 * An implementation of the model object '<em><b>Serial_Ext_GPS_NMEA0183</b></em>'.
 * @generated
 */
public class Serial_Ext_GPS_NMEA0183 extends TopicObject implements ISerial_Ext_GPS_NMEA0183 {
	protected EList<PropertyChangeListener> listeners;
	protected IGPS_NMEA0183_T_R serial_Ext_GPS_NMEA0183_Data;

	/**
	 * @generated
	 */
	protected Serial_Ext_GPS_NMEA0183() {
		super();
		initInterfaceDetails();
	}

	private void initInterfaceDetails() {
		interfaceDetailMap.put("fek/9465", new InterfaceDetails("fek/9465"));
	}

	/**
	 * @generated
	 */
	@Override
	protected EClass eStaticClass() {
		return Serial_ext_gps_nmea0183Package.Literals.SERIAL_EXT_GPS_NMEA0183;
	}

	/**
	 * @generated
	 */
	public IGPS_NMEA0183_T_R getSerial_Ext_GPS_NMEA0183_Data() {
		if (serial_Ext_GPS_NMEA0183_Data != null && serial_Ext_GPS_NMEA0183_Data.eIsProxy()) {
			InternalEObject oldSerial_Ext_GPS_NMEA0183_Data = (InternalEObject) serial_Ext_GPS_NMEA0183_Data;
			serial_Ext_GPS_NMEA0183_Data = (IGPS_NMEA0183_T_R) eResolveProxy(oldSerial_Ext_GPS_NMEA0183_Data);
			if (serial_Ext_GPS_NMEA0183_Data != oldSerial_Ext_GPS_NMEA0183_Data) {
			}
		}
		return serial_Ext_GPS_NMEA0183_Data;
	}

	/**
	 * @generated
	 */
	public IGPS_NMEA0183_T_R basicGetSerial_Ext_GPS_NMEA0183_Data() {
		return serial_Ext_GPS_NMEA0183_Data;
	}

	/**
	 * @generated
	 */
	public void setSerial_Ext_GPS_NMEA0183_Data(IGPS_NMEA0183_T_R newSerial_Ext_GPS_NMEA0183_Data) {
		IGPS_NMEA0183_T_R oldSerial_Ext_GPS_NMEA0183_Data = serial_Ext_GPS_NMEA0183_Data;
		serial_Ext_GPS_NMEA0183_Data = newSerial_Ext_GPS_NMEA0183_Data;
		FCALLogs.getInstance().log
				.debug(LoggerConstants.LOG_SDK_PREFIX + " Serial_Ext_GPS_NMEA0183 property \"serial_Ext_GPS_NMEA0183_Data\" changed from  "
						+ oldSerial_Ext_GPS_NMEA0183_Data + " to " + newSerial_Ext_GPS_NMEA0183_Data);
		notifyPropertyChange("serial_Ext_GPS_NMEA0183_Data", oldSerial_Ext_GPS_NMEA0183_Data, newSerial_Ext_GPS_NMEA0183_Data);
	}

	/**
	 * @generated
	 */
	public long getSerial_Ext_GPS_NMEA0183_Data_Timestamp() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9465");
		if (interfaceDetails != null) {
			return interfaceDetails.getTimestmap();
		}
		return -1;
	}

	/**
	 * @generated
	 */
	public boolean isSerial_Ext_GPS_NMEA0183_Data_Valid() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9465");
		if (interfaceDetails != null) {
			return interfaceDetails.isValid();
		}
		return false;
	}

	/**
	 * @generated
	 */
	public void addPropertyChangeListener(PropertyChangeListener listener) {
		if (listeners == null) {
			listeners = new BasicEList<>();
		}
		listeners.add(listener);
	}

	/**
	 * @generated
	 */
	public void removePropertyChangeListener(PropertyChangeListener listener) {
		if (listeners != null) {
			listeners.remove(listener);
		}
	}

	/**
	 * @generated
	 */
	public void notifyPropertyChange(String name, Object oldValue, Object newValue) {
		if (listeners != null) {
			for (PropertyChangeListener listener : this.listeners) {
				listener.propertyChange(new PropertyChangeEvent(this, name, oldValue, newValue));
			}
		}
	}

	/**
	 * @generated
	 */
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
		case Serial_ext_gps_nmea0183Package.SERIAL_EXT_GPS_NMEA0183__SERIAL_EXT_GPS_NMEA0183_DATA:
			if (resolve)
				return getSerial_Ext_GPS_NMEA0183_Data();
			return basicGetSerial_Ext_GPS_NMEA0183_Data();
		}
		return super.eGet(featureID, resolve, coreType);
	}

	/**
	 * @generated
	 */
	@SuppressWarnings("unchecked")
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
		case Serial_ext_gps_nmea0183Package.SERIAL_EXT_GPS_NMEA0183__SERIAL_EXT_GPS_NMEA0183_DATA:
			setSerial_Ext_GPS_NMEA0183_Data((IGPS_NMEA0183_T_R) newValue);
			return;
		}
		super.eSet(featureID, newValue);
	}

	/**
	 * @generated
	 */
	@Override
	public void eUnset(int featureID) {
		switch (featureID) {
		case Serial_ext_gps_nmea0183Package.SERIAL_EXT_GPS_NMEA0183__SERIAL_EXT_GPS_NMEA0183_DATA:
			setSerial_Ext_GPS_NMEA0183_Data((IGPS_NMEA0183_T_R) null);
			return;
		}
		super.eUnset(featureID);
	}

	/**
	 * @generated
	 */
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
		case Serial_ext_gps_nmea0183Package.SERIAL_EXT_GPS_NMEA0183__SERIAL_EXT_GPS_NMEA0183_DATA:
			return serial_Ext_GPS_NMEA0183_Data != null;
		}
		return super.eIsSet(featureID);
	}
} //Serial_Ext_GPS_NMEA0183

File: master_of_plow_CPP_SDK/src-gen/nevonex/serial_ext_gps_nmea0183/Serial_Ext_GPS_NMEA0183.cpp
/*
 * nevonex/serial_ext_gps_nmea0183/Serial_Ext_GPS_NMEA0183.cpp
 * Copyright (c) Robert Bosch GmbH. All rights reserved.
 */

#include "Serial_Ext_GPS_NMEA0183.hpp"
#include <ecore/EObject.hpp>
#include <ecore/EClass.hpp>
#include "nevonex/serial_ext_gps_nmea0183/Serial_ext_gps_nmea0183Package.hpp"
#include <ecorecpp/mapping.hpp>
#include <nevonex-fcal-platform/log/Logger.hpp>

/*PROTECTED REGION ID(Serial_Ext_GPS_NMEA0183.cpp) START*/
// Please, enable the protected region if you add manually written code.
// To do this, add the keyword ENABLED before START.
/*PROTECTED REGION END*/

using namespace ::nevonex::serial_ext_gps_nmea0183;
using namespace ::nevonex::log;

// Default constructor
Serial_Ext_GPS_NMEA0183::Serial_Ext_GPS_NMEA0183() : m_serial_Ext_GPS_NMEA0183_Data(0)
{
    // Machine does NOT call _initialize() — Provider manages initialization

    m_interfaceDetailMap =
    {
    { 9465, ::nevonex::types::InterfaceDetails("fek/9465") } };

    /*PROTECTED REGION ID(Serial_Ext_GPS_NMEA0183__Serial_Ext_GPS_NMEA0183) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

Serial_Ext_GPS_NMEA0183::~Serial_Ext_GPS_NMEA0183()
{
    /*PROTECTED REGION ID(Serial_Ext_GPS_NMEA0183__Destructor) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

// References
::nevonex::fcal::GPS_NMEA0183_T_R_ptr Serial_Ext_GPS_NMEA0183::getSerial_Ext_GPS_NMEA0183_Data() const
{
    return m_serial_Ext_GPS_NMEA0183_Data;
}

void Serial_Ext_GPS_NMEA0183::setSerial_Ext_GPS_NMEA0183_Data(::nevonex::fcal::GPS_NMEA0183_T_R_ptr _serial_Ext_GPS_NMEA0183_Data)
{
    ::nevonex::fcal::GPS_NMEA0183_T_R_ptr _old_serial_Ext_GPS_NMEA0183_Data = m_serial_Ext_GPS_NMEA0183_Data;
    m_serial_Ext_GPS_NMEA0183_Data = _serial_Ext_GPS_NMEA0183_Data;
    SDK_FCAL_LOG(SeverityLevel::debug) << "Serial_Ext_GPS_NMEA0183 property \"serial_Ext_GPS_NMEA0183_Data\" has changed.";
    notifyPropertyChange("serial_Ext_GPS_NMEA0183_Data", _old_serial_Ext_GPS_NMEA0183_Data, m_serial_Ext_GPS_NMEA0183_Data);
}

// FCAL timestamp/valid
::ecore::ELong Serial_Ext_GPS_NMEA0183::getSerial_Ext_GPS_NMEA0183_Data_Timestamp()
{
    auto _detail = m_interfaceDetailMap.find(9465);
    if (_detail == m_interfaceDetailMap.end())
        return -1;
    return _detail->second.getTimestamp();
}

::ecore::EBoolean Serial_Ext_GPS_NMEA0183::isSerial_Ext_GPS_NMEA0183_Data_Valid()
{
    auto _detail = m_interfaceDetailMap.find(9465);
    if (_detail == m_interfaceDetailMap.end())
        return false;
    return _detail->second.isValid();
}

::ecore::EBoolean Serial_Ext_GPS_NMEA0183::updateInterfaceDetail(int _topicKey, ::ecore::EBoolean _valid, ::ecore::ELong _timestamp)
{
    auto _detail = m_interfaceDetailMap.find(_topicKey);
    if (_detail == m_interfaceDetailMap.end())
    {
        SDK_FCAL_LOG(SeverityLevel::warning) << _topicKey
                << " interface details are not present in the interface detail map[Serial_Ext_GPS_NMEA0183].";
        return false;
    }

    if (_detail->second.getTimestamp() > _timestamp)
    {
        SDK_FCAL_LOG(SeverityLevel::warning)
                << "Old or duplicate message received. Old TS : "
                << _detail->second.getTimestamp() << " new TS : " << _timestamp
                << " feature: Serial_Ext_GPS_NMEA0183 topic:" << _topicKey;
        return false;
    }

    _detail->second.setValid(_valid, _timestamp);
    return true;
}

// PropertyChange delegation (diamond inheritance requires explicit override)
void Serial_Ext_GPS_NMEA0183::addPropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::addPropertyChangeListener(_listener);
}

void Serial_Ext_GPS_NMEA0183::removePropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::removePropertyChangeListener(_listener);
}

void Serial_Ext_GPS_NMEA0183::notifyPropertyChange(::ecore::EString const &_name,
        ::ecore::EJavaObject const &_oldValue,
        ::ecore::EJavaObject const &_newValue)
{
    using namespace ::nevonex::types;
    PropertyChange::notifyPropertyChange(_name, _oldValue, _newValue);
}
